### DISCUSSION TITLE
Entanglement Purification Rule Action

### DISCUSSION THREAD

**Discussion by LesterKim** [sequence-toolbox/SeQUeNCe#71] [Q&A]
In the requirement function for entanglement purification, what is the significance of [the following lines of code](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L299-L308)?

My interpretation is that `memories_info[0].remote_node` has two purification protocols, one that has a kept memory entangled with `memories_info[0].memory` and another that has a kept memory entangled with `memories_info[1].memory`. Those two protocols are saved in `_protocols`. But then, why do you set `_protocols[0].meas_memo = _protocols[1].kept_memo`? I don't understand the reasoning in [the lines highlighted](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L299-L308).

For context, I am wondering how to add a purification protocol that sacrifices two measured entangled pairs to purify the kept memory (e.g. single-selection double-error purification).

**Comment by xwu64** [comment]
Although I cannot open attached links, I assume the code about your questions is at [here](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L302). 

The logic of rules for purification protocols involves two routers to cooperate. Two routers perform different operations on the entangled memories. Let's name two routers to router A and router B here. Router A's action of rule is to [select two qualified memories for the purification protocol](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L282). It's impossible to make router B perform the same action without information from router A. For example, if router A and B have three pairs of qualified memories (a1-b1, a2-b2, a3-c3) for purification, the action on router A may select a1-b1 and a2-b2 and the same action on router B may select a1-b1 and a3-b3, which prohibits cooperations of protocols for purification. 

To avoid such situation, the action of router B is to [create an instance of entanglement purification protocol to hold memory temporarily](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L333). Since the purification protocol only holds one memory, the protocol will not be executed. When the message from router A is received by router B, the message will notify router B which memories are selected for one purification protocol. The function ["req_func"](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L285) will scan these incomplete purification protocols to select incomplete  protocols that hold corresponding memories. The "req_func" will use these two protocols to assemble a complete protocol to cooperate with the protocol on router A. 

The referred code lines is to assemble a complete protocol. `_protocols[0]` denotes the protocol that holds the kept memory. `_protocols[1]` denotes the protocol that holds the measured memory. After the assembly,   `_protocols[0]` is used to purify entanglement and `_protocols[1]` is discarded. 

I know this implementation may not very eligible. I'm welcome ideas about improving the quality of code at here.

**Reply by LesterKim** [reply]
Thank you so much for the response. I update the attached links, so I apologize for that. Your explanation makes sense, and I will think about ideas for improving the quality of code here. It will require some refactoring, so I will try to get some in this month.

**Reply by LesterKim** [reply]
When you create an instance of entanglement purification protocol to hold memory temporarily in router B, [why do you only select `memories[0]`](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L335-L336)? Do you want router B to create an instance for each its entangled memories?

Again, for context, I am thinking about how to support more purification protocols that sacrifice two or more entangled pairs to purify an a `kept_memo`.

**Reply by LesterKim** [reply]
I see that the answer to my question is that [the rule condition returns a list with only one memory info](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L324). Is that sufficient because the rule condition will be applied to the memory info for each memory in the node?

**Reply by xwu64** [reply]
I don't get your question. What will insufficient? 

For the purification protocols that consumes multiple memories, the rule need to be designed specific for the protocol. I'm thinking moving the logic of condition and action to the class of protocols may be more reasonable. Rule manager could call functions from these entanglement protocols without knowledge about the requirement of protocols.

**Reply by LesterKim** [reply]
Since the rule condition returns a list of only one memory info, is [that rule condition (and action)](https://github.com/sequence-toolbox/SeQUeNCe/blob/8cbcbd31f013e511d5cbaf96399711161ddadabe/src/network_management/reservation.py#L321-L337) applied to every memory in the node during a simulation? After all, we need at least two entangled memories in the node to run purification.

**Reply by LesterKim** [reply]
I am reading [chapter 4 of the tutorial](https://github.com/sequence-toolbox/SeQUeNCe/blob/master/docs/source/tutorial/chapter4/resource_management.md) again. So the rule condition is applied to all the memories in the node in a given range.

**Reply by xwu64** [reply]
Sorry for responding late. You're right the rule will check all memories to search qualified memories.
